### ISSUE TITLE
Challenges of end-to-end machine learning with SQLFlow

### ISSUE THREAD

**Issue by wangkuiyi** [sql-machine-learning/sqlflow#2277]
For 

- TensorFlow
- PyTorch
- XGBoost

we need to consider the following questions

1. How to combine the in-train-loop data transformation and the training iteration, and how to make the combination customizable by the `COLUMN` clause.

   - It seems that XGBoost doesn't support customize the train loop.   We might need to invent something new.

1. We hope the train-loop works with distributed training as well.

   - ElasticDL is a good example. It provides a train-loop, that calls data transformation logic derived from the COLUMN-clause and the train iteration.
   - If it is too challenging to make the train-loop runnable in both local and distributed mode, let us consider only the distributed mode for Kubernetes.

1. We hope the in-train-loop data transformation logic reusable in both training and (online) prediction.

**Comment by typhoonzero** [comment]
The current design for support XGBoost training with a string column: https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/design/figures/xgboost_feature_column.md

Support this in XGBoost distributed training is simple: just apply the same transform Python functions to generate DMatrix file on each worker.

**Comment by brightcoder01** [comment]
> The current design for support XGBoost training with a string column: https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/design/figures/xgboost_feature_column.md
> 
> Support this in XGBoost distributed training is simple: just apply the same transform Python functions to generate DMatrix file on each worker.

Do this transformation/preprocessing while generating the DMatrix file is a feasible option.
The key point we need figure out is how we can combine the preprocessing logic for DMatrix generation and XGBoost model together into PMML for online prediction.

**Comment by sneaxiy** [comment]
@brightcoder01 I agree. We should rethink the online serving cases mentioned in https://github.com/sql-machine-learning/sqlflow/issues/2190 . The design in   https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/design/xgboost_feature_column.md is easy for offline training and prediction, but not online serving.

**Comment by brightcoder01** [comment]
For PyTorch preprocessing, the investigations are concluded in #2276 . The key challenge for PyTorch is that how to save the preprocess logic together with serialized PyTorch model.
